pkg/apis/serving/v1/revision_helpers.go

```diff
@@ -87,6 +107,41 @@ func (rs *RevisionSpec) GetContainer() *corev1.Container {
 	return &corev1.Container{}
 }
 
+// SetRoutingState sets the routingState label on this Revision and updates the
+// routingStateModified annotation.
+func (r *Revision) SetRoutingState(state RoutingState) {
+	stateStr := string(state)
+	if t := r.ObjectMeta.Annotations[serving.RoutingStateModifiedAnnotationKey]; t != "" &&
+		r.Labels[serving.RoutingStateLabelKey] == stateStr {
+		return // Don't update timestamp if no change.
+	}
+
+	if r.Labels == nil {
+		r.Labels = make(map[string]string, 1)
+	}
+	r.Labels[serving.RoutingStateLabelKey] = stateStr
+
+	if r.Annotations == nil {
+		r.Annotations = make(map[string]string, 1)
+	}
+	r.Annotations[serving.RoutingStateModifiedAnnotationKey] = time.Now().Format(time.RFC3339)
+}
+
+// GetRoutingState retrieves the RoutingState label.
+func (r *Revision) GetRoutingState() RoutingState {
+	return RoutingState(r.ObjectMeta.Labels[serving.RoutingStateLabelKey])
+}
+
+// GetRoutingStateModified retrieves the RoutingStateModified annotation.
+func (r *Revision) GetRoutingStateModified() time.Time {
+	val := r.ObjectMeta.Annotations[serving.RoutingStateModifiedAnnotationKey]
+	parsed, err := time.Parse(time.RFC3339, val)
+	if err != nil {
+		return time.Time{}
```

Comment:
Is there value in distinguishing when it's set or not? ie. all unmarked revisions would have the epoch (empty) time

Reply:
I'd prefer not to, default value should be fine to compare to. You can see what was done with GetLastPinned() and it adds a lot of complexity to have a not-present error value that's used later. I'd prefer to add it later if we find value in it (but I don't expect to be setting epoch anyway)